### PR TITLE
Add Node.js 6 runtime

### PR DESCRIPTION
**Pull request by jasonpet**
PR for issue #17

### CODE REVIEW

**Comment by jasonpet** [comment]
@rabbah I see the following output in my logs for nodejs 6: "XXX_THE_END_OF_A_WHISK_ACTIVATION_XXX"

It appears that the invoker is looking for this in the existing nodejs and is filtering this out.  Should I be updating the code to handle this for nodejs 6 as part of this PR or is thing something that will be handled separately?

**Comment by perryibm** [comment]
@paulcastro 

Hi Jason.  The invoker needs the sentinel markers to be coordinated with the language container and since this is node (which already has this feature), it should be enabled as part of this PR.  We are planning on doing this along with Swift and Python.  The relevant code is in lines 295 and 311 of Invoker.scala.

On line 295, augment with
```
private val nodejs6ImageName = WhiskAction.containerImageName(NodeJS6Exec("", None), config.dockerRegistry, config.dockerImageTag)
```

and change line 311 to
```
val isNodeJs = con.image == nodejsImageName || con.image == nodejs6ImageName
```

That should work but a better version would be to create a Set of all image names that will have the sentinel marker and change to checking `con.image` for membership in such a set.

**Comment by jasonpet** [comment]
I will update the nodejs 6 container to contain the two new packages being added as part of issue #523.   I assume docs will need to be updated as well to include the nodejs 6 list of packages.  The list is currently the same right now outside of package versions (which I don't know if we even list) and Jade being renamed to Pug.

**Comment by jasonpet** [comment]
pg #327

**Comment by psuter** [comment]
The changes look like they will work. I am a little uncomfortable though with the approach we are taking:

    $ find nodejsAction/ -type f -exec wc -l {} + | tail -1
    925 total
    $ find nodejs6Action/ -type f -exec wc -l {} + | tail -1
    916 total
    $ diff -rw nodejsAction/ nodejs6Action/ | wc -l
    104

We are copy-pasting a non-trivial set of functionalities (container definition, action wrapper, Whisk "SDK"), with basically no changes except pinning version numbers of the runtime and of the packages. I know we've recently had similar instances of clone-and-own, but never to that extent. Can we refactor this into `core/nodejsCommon` containing almost all the code, and `core/nodejs[6]Action/` using a common base `Dockerfile`? I feel like this is doomed to diverge in unpleasant ways otherwise.

@sjfink @rabbah and others may want to comment and/or suggest better refactoring strategies.

**Comment by perryibm** [comment]
Agreed we should not have multiple copies of all those js files.  It will be a maintenance nightmare.  Your strategy seems fine.  Since docker likes to build things under it, having a separate Dockerfile to build a base is necessary as that Dockerfile will have access to the shared stuff.  And of course, faster builds will be possible if we do it this way.

**Comment by jasonpet** [comment]
Seems like some of this will diverge very soon though with issues like #78 and #505.

**Comment by rabbah** [comment]
Clone and own is bad. Extend and override is good.

**Comment by jasonpet** [comment]
@psuter - I pushed my refactoring changes and ran pg build # 335

**Comment by psuter** [comment]
Code looks good to me. Should we have at least one test that's specific to Node.js 6? Maybe throw in a `let` somewhere?

**Comment by rabbah** [comment]
Can the packages be shared too? I see image magik in the base dockerhost why not all the other packages?

This is borderline nitpicking: the base could have just been the current nodejs action and then the specialization is node4 vs node6. To make the version more explicit (where node4's name is what the system current uses for its kind).

**Comment by jasonpet** [comment]
@psuter - I added a test for testing default function parameters, which is new to node 6.  I threw in the `let` you mentioned as well which not new to node 6, but isn't supported in the existing node we have (0.12.14).  My idea to throw in the Promises to test node 6 was to no avail since those are even supported in our existing node.  Found this nice link for viewing node features by version: http://node.green/

**Comment by rabbah** [comment]
@jasonpet if we kept nodejsAction as is including its docker file, could the nodejs6 action Dockerfile simply upgrade the node version using whisk/nodejsaction as the image base?

the thinking is that we have only one list of packages to maintain instead of two (when we diverge, we'll diverge and make the lists explicit).

File: tests/src/packages/UtilsTests.scala [hunk 1]
```diff
@@ -72,26 +91,45 @@ class UtilsTests extends TestHelpers with WskTestHelpers with Matchers {
...
+      */
+    it should "split a string into an array of strings using the split action on nodejs 6" in withAssetCleaner(wskprops) {
+        (wp, assetHelper) =>
+            val wsk = new Wsk()
+
+            val file = TestUtils.getCatalogFilename("utils/split.js")
+            val actionName = "splitNodejs6"
+
+            assetHelper.withCleaner(wsk.action, actionName) { (action, _) =>
+                action.create(name = actionName, artifact = Some(file), kind = Some("nodejs:6"), expectedExitCode = 0)
```
Comment (rabbah) [inline tests/src/packages/UtilsTests.scala]: `expectedExitCode = 0` is the default and may be omitted.

**Comment by jasonpet** [comment]
@rabbah - I haven't really been able to follow your last few comments regarding the docker file split.  The npm packages are different then the docker packages.  image magik is not an npm package.  The npm packages have different versions based on the node version and even different names in some cases.  Jade was renamed to Pug.

**Comment by rabbah** [comment]
I'm thinking if there is a way to reduce the duplication between the two dockerfiles: primarily so we aren't maintaining two package lists. I did notice the versions are different, but it wasn't clear that we couldn't upgrade the packages in both runtimes. You are right that if we did this, then some packages will have breaking changes for existing node actions, so perhaps this is a reason not to consolidate further.

**Comment by jasonpet** [comment]
@psuter - I believe this should be ready now.  I had run pg build 338 after making those test updates.

**Comment by psuter** [comment]
It's [in](04b6268895001aa556e6d579f4c3e0a5a9901a04).

**Comment by markusthoemmes** [comment]
We should extend that split to all our actioncontainers maybe? Only having it in nodejs feels like adding a lot of maintainability and build-process overhead to just installing one package in the same image? Shouldn't stuff like "imagemagick" be in every actionContainer we have?

**Pull request merged**